File: EPSILON/TOOLS/Altova/Altova/XMLSpy2006/spl/java/altova/text/edifact/Scanner.java
[call GenerateFileHeader("Scanner.java")]
package com.altova.text.edifact;

import java.util.ArrayList;
import java.util.Iterator;
import java.lang.StringBuffer;

public class Scanner {
	public class SeparatorToSymbolicNameMap {
		private class SeparatorSymbolicNamePair {
			char m_Separator = 0;

			String m_SymbolicName = "";

			public SeparatorSymbolicNamePair(char separator, String symbolicname) {
				m_Separator = separator;
				m_SymbolicName = symbolicname;
			}

			public char getSeparator() {
				return m_Separator;
			}

			public String getSymbolicName() {
				return m_SymbolicName;
			}
		}

		ArrayList m_List = new ArrayList();

		boolean isValueContained(String symbolicname) {
			for (int i = 0; i < m_List.size(); ++i) {
				SeparatorSymbolicNamePair entry = (SeparatorSymbolicNamePair) m_List
						.get(i);
				if (symbolicname == entry.getSymbolicName())
					return true;
			}
			return false;
		}

		void removeAllEntriesWithValue(String symbolicname) {
			Iterator it = m_List.iterator();
			while (it.hasNext()) {
				SeparatorSymbolicNamePair entry = (SeparatorSymbolicNamePair) it
						.next();
				if (symbolicname == entry.getSymbolicName())
					it.remove();
			}
		}

		public void add(char separator, String symbolicname) {
			if (this.isValueContained(symbolicname))
				this.removeAllEntriesWithValue(symbolicname);
			SeparatorSymbolicNamePair entry = new SeparatorSymbolicNamePair(
					separator, symbolicname);
			m_List.add(entry);
		}

		public String getSymbolicNameForSeparator(char separator) {
			for (int i = 0; i < m_List.size(); ++i) {
				SeparatorSymbolicNamePair entry = (SeparatorSymbolicNamePair) m_List
						.get(i);
				if (separator == entry.getSeparator())
					return entry.getSymbolicName();
			}
			return null;
		}

		public char getSeparatorForSymbolicName(String name) {
			for (int i = 0; i < m_List.size(); ++i) {
				SeparatorSymbolicNamePair entry = (SeparatorSymbolicNamePair) m_List
						.get(i);
				if (name == entry.getSymbolicName())
					return entry.getSeparator();
			}
			return 0;
		}

		public boolean isContained(char separator) {
			return (null != this.getSymbolicNameForSeparator(separator));
		}
	}

	private String m_Current;

	private int m_Position;

	private String m_Token;

	private String m_SeparatorToken;

	private char m_EscapeChar;

	private char m_DecimalSeparator;

	private SeparatorToSymbolicNameMap m_SeparatorMap = new SeparatorToSymbolicNameMap();

	private StringBuffer m_TokenBuffer = new StringBuffer();

	private void moveNext() {
		++m_Position;
	}

	private void moveBack() {
		--m_Position;
	}

	private boolean isDelimiter() {
		return m_SeparatorMap.isContained(this.getCurrentCharacter());
	}

	private boolean isInsideExpression() {
		if (this.isEndOfText())
			return false;

		if (m_EscapeChar == this.getCurrentCharacter()) {
			this.moveNext();
			return true;
		}

		return !this.isDelimiter();
	}

	private short getScannerReturnCode(boolean checktoken) {
		if (this.isEndOfText())
			return END;
		else {
			if ((checktoken) && (0 == m_Token.length()))
				return EMPTY;
			else
				return VALUE;
		}
	}

	final public static short VALUE = 0;

	final public static short EMPTY = 1;

	final public static short END = 2;

	public Scanner() {
		m_SeparatorMap.add('\\0', "terminate");
		m_SeparatorMap.add('\\r', "return");
		m_SeparatorMap.add('\\n', "linefeed");
		m_EscapeChar = '\\0';
	}

	public void init(String buffer) {
		m_Current = buffer;
		m_Position = 0;
	}

	public String getToken() {
		return m_Token;
	}

	public short scanExpression() {
		if (this.isEndOfText())
			return END;

		m_TokenBuffer.delete(0, m_TokenBuffer.length());

		if (this.isInsideExpression())
			m_TokenBuffer.append(getCurrentCharacter());

		this.moveNext();
		while (this.isInsideExpression()) {
			m_TokenBuffer.append(getCurrentCharacter());
			this.moveNext();
		}

		m_Token = m_TokenBuffer.toString();

		// store the current separator token so that we can depend on it in a rule.
		m_SeparatorToken = m_SeparatorMap.getSymbolicNameForSeparator(getCurrentCharacter());

		return this.getScannerReturnCode(true);
	}

	public short scanOneCharacter() {
		this.moveNext();
		m_Token = "";
		m_Token += this.getCurrentCharacter();
		return this.getScannerReturnCode(false);
	}

	public short moveBackOneCharacter() {
		moveBack();
		m_Token = "";
		m_Token += this.getCurrentCharacter();
		return this.getScannerReturnCode(false);
	}

	public char getCurrentCharacter() {
		if (m_Position >= m_Current.length())
			return '\\0';
		return m_Current.charAt(m_Position);
	}

	public SeparatorToSymbolicNameMap getSeparatorMap() {
		return m_SeparatorMap;
	}

	public String getSeparatorToken() {
		return m_SeparatorToken;
	}

	public boolean isEndOfText() {
		return (0 == getCurrentCharacter());
	}

	public void setEscapeChar(char rhs) {
		m_EscapeChar = rhs;
	}

	public char getDecimalSeparator() {
		return m_DecimalSeparator;
	}

	public void setDecimalSeparator(char rhs) {
		m_DecimalSeparator = rhs;
	}
}

File: EPSILON/SAMPLE FILE/E-publishing/Tools/Altova/Altova/XMLSpy2006/spl/cs/AltovaText/EDIFact/Scanner.cs
[call GenerateFileHeader("Scanner.cs")]
using System;
using System.Collections;
using System.Text;

namespace Altova.TextParser.EDIFACT
{
	/// <summary>
	/// Encapsulates scanning a string buffer for tokens and delimiters.
	/// </summary>
	public class Scanner
	{
		#region Implementation Detail:
		string mCurrent; // complete buffer
		int mPosition; // current position
		string mToken; // current token value
		string mSeparatorToken; // current separator
		char mEscapeChar = '\\0';
		char mQuoteChar = '\\0';
		char mDecimalSeparator = ',';
		Hashtable mSeparatorMap = new Hashtable();
		Hashtable mTokenMap = new Hashtable();
		string mTokenizerToken = "";
		StringBuilder mTokenBuilder = new StringBuilder();

		void MoveNext()
		{
			++mPosition;
		}
		void MoveBack()
		{
			--mPosition;
		}
		void MapToken(string rhs)
		{
			if (mTokenMap.ContainsKey(rhs))
				mTokenizerToken = (string) mTokenMap\[rhs\];
		}
		bool IsInsideExpression
		{
			get
			{
				if (this.IsEndOfText) return false;
				if (mEscapeChar == this.CurrentChar)
				{
					this.MoveNext();
					return true;
				}
				return !mSeparatorMap.ContainsKey(this.CurrentChar);
			}
		}
		void EraseAllSeparatorMappingsForString(string name)
		{
			string keys = "";
			foreach (DictionaryEntry entry in mSeparatorMap)
			{
				if (name == (string) entry.Value)
					keys += (char) entry.Key;
			}
			foreach (char c in keys)
				mSeparatorMap.Remove(c);
		}
		#endregion
		#region Nested Types:
		/// <summary>
		/// Encapsulates the possible results of a call to <see cref="Scanner.Scan"/>().
		/// </summary>
		public enum Indicator
		{
			/// <summary>a token has been found</summary>
			Value,
			/// <summary>an empty token has been found</summary>
			Empty,
			/// <summary>the end of the buffer has been encountered</summary>
			End,
		}
		#endregion
		#region Public Interface:
		/// <summary>
		/// Constructs an instance of this class.
		/// </summary>
		public Scanner()
		{
			this.AddSeparator('\\0', "terminate");
			/*
			this.AddSeparator('\\r', "return");
			this.AddSeparator('\\n', "linefeed");
			*/
		}
		/// <summary>
		/// Initializes this instance for scanning a new buffer.
		/// </summary>
		/// <param name="buffer">the new buffer to scan</param>
		public void Init(string buffer)
		{
			mCurrent = buffer;
			mPosition = 0;
		}
		/// <summary>
		/// Gets the current token.
		/// </summary>
		public string Token
		{
			get
			{
				return mToken;
			}
		}
		/// <summary>
		/// Gets the current value - alias for <see cref="Token"/>.
		/// </summary>
		public string Value
		{
			get
			{
				return mToken;
			}
		}
		/// <summary>
		/// Scans forward inside the buffer until a complete token can be mapped or
		/// a separator is found.
		/// </summary>
		/// <returns>a value indicating what has been found</returns>
		public Indicator Scan()
		{
			if (this.IsEndOfText) return Indicator.End;

			mTokenBuilder.Remove(0, mTokenBuilder.Length);

			if (this.IsInsideExpression)
				mTokenBuilder.Append(this.CurrentChar);
			for (this.MoveNext(); this.IsInsideExpression; this.MoveNext())
				mTokenBuilder.Append(this.CurrentChar);
			mToken = mTokenBuilder.ToString();
			this.MapToken(mToken);

			// store the current separator token so that we can depend on it in a rule.
			mSeparatorToken = this.Separator;

			if (this.IsEndOfText)
				return Indicator.End;
			else if (0 == mToken.Length)
				return Indicator.Empty;
			else
				return Indicator.Value;
		}
		/// <summary>
		/// Scans ahead one character.
		/// </summary>
		/// <returns>a value indicating what has been found</returns>
		public Indicator ScanChar()
		{
			this.MoveNext();
			mToken = string.Empty;
			mToken += this.CurrentChar;
			return this.IsEndOfText ? Indicator.End : Indicator.Value;
		}
		/// <summary>
		/// Scans back one character.
		/// </summary>
		/// <returns>a value indicating what has been found</returns>
		public Indicator BackChar()
		{
			this.MoveBack();
			mToken = string.Empty;
			mToken += this.CurrentChar;
			return this.IsEndOfText ? Indicator.End : Indicator.Value;
		}
		/// <summary>
		/// Gets the current char in the buffer.
		/// </summary>
		public char CurrentChar
		{
			get
			{
				if (mPosition >= mCurrent.Length) return '\\0';
				return mCurrent\[mPosition\];
			}
		}
		/// <summary>
		/// Gets the separator for the current char.
		/// </summary>
		public string Separator
		{
			get
			{
				return (string) mSeparatorMap\[this.CurrentChar\];
			}
		}
		/// <summary>
		/// Gets the current separator token.
		/// </summary>
		public string SeparatorToken
		{
			get
			{
				return mSeparatorToken;
			}
		}
		/// <summary>
		/// Gets whether scanning has arrived at the end of the buffer.
		/// </summary>
		public bool IsEndOfText
		{
			get
			{
				return (mPosition >= mCurrent.Length);
			}
		}
		/// <summary>
		/// Maps a character to a separator token.
		/// </summary>
		/// <param name="ch">the character</param>
		/// <param name="name">the separator token</param>
		/// <remarks>
		/// If the separator token is already mapped to some character,
		/// all these existing mappings will be erased.
		/// </remarks>
		public void AddSeparator(char ch, string name)
		{
			this.EraseAllSeparatorMappingsForString(name);
			mSeparatorMap\[ch\] = name;
		}
		/// <summary>
		/// Gets a separator by its symbolic name.
		/// </summary>
		/// <param name="name">the symbolic name</param>
		/// <returns>the separator</returns>
		public char GetSeparatorByName(string name)
		{
			foreach (DictionaryEntry entry in mSeparatorMap)
				if (entry.Value.ToString()==name)
					return (char) entry.Key;
			return '\\0';
		}
		/// <summary>
		/// Get/sets the escape character used by this instance.
		/// </summary>
		public char EscapeChar
		{
			get
			{
				return mEscapeChar;
			}
			set
			{
				mEscapeChar = value;
			}
		}
		/// <summary>
		/// Get/sets the quote character used by this instance.
		/// </summary>
		public char QuoteChar
		{
			get
			{
				return mQuoteChar;
			}
			set
			{
				mQuoteChar = value;
			}
		}
		/// <summary>
		/// Get/sets the character used by instance as decimal separator.
		/// </summary>
		public char DecimalSeparator
		{
			get
			{
				return mDecimalSeparator;
			}
			set
			{
				mDecimalSeparator = value;
			}
		}
		#endregion
	}
}